tests/ui/type/mutually-recursive-types.rs

```diff
@@ -1,15 +1,54 @@
+//! Test that mutually recursive type definitions are properly handled by the compiler.
+//! This checks that types can reference each other in their definitions through
+//! `Box` indirection, creating cycles in the type dependency graph.
+
 //@ run-pass
 
-#![allow(non_camel_case_types)]
-#![allow(dead_code)]
+#[derive(Debug, PartialEq)]
+enum Colour {
+    Red,
+    Green,
+    Blue,
+}
+
+#[derive(Debug, PartialEq)]
+enum Tree {
+    Children(Box<List>),
+    Leaf(Colour),
+}
+
+#[derive(Debug, PartialEq)]
+enum List {
+    Cons(Box<Tree>, Box<List>),
+    Nil,
+}
+
+#[derive(Debug, PartialEq)]
+enum SmallList {
+    Kons(isize, Box<SmallList>),
+    Neel,
+}
+
+pub fn main() {
+    // Construct and test all variants of Colour
+    let leaf_red = Tree::Leaf(Colour::Red);
+    assert_eq!(leaf_red, Tree::Leaf(Colour::Red));
 
+    let leaf_green = Tree::Leaf(Colour::Green);
+    assert_eq!(leaf_green, Tree::Leaf(Colour::Green));
 
-enum colour { red, green, blue, }
+    let leaf_blue = Tree::Leaf(Colour::Blue);
+    assert_eq!(leaf_blue, Tree::Leaf(Colour::Blue));
 
-enum tree { children(Box<list>), leaf(colour), }
+    let empty_list = List::Nil;
+    assert_eq!(empty_list, List::Nil);
 
-enum list { cons(Box<tree>, Box<list>), nil, }
+    let tree_with_children = Tree::Children(Box::new(List::Nil));
+    assert_eq!(tree_with_children, Tree::Children(Box::new(List::Nil)));
 
-enum small_list { kons(isize, Box<small_list>), neel, }
+    let list_with_tree = List::Cons(Box::new(Tree::Leaf(Colour::Blue)), Box::new(List::Nil));
+    assert_eq!(list_with_tree, List::Cons(Box::new(Tree::Leaf(Colour::Blue)), Box::new(List::Nil)));
 
-pub fn main() { }
+    let small_list = SmallList::Kons(42, Box::new(SmallList::Neel));
+    assert_eq!(small_list, SmallList::Kons(42, Box::new(SmallList::Neel)));
+}
```

Comment:
I think these new assertions should be dropped; this is a type system UI test, the test cases asserting `x == x` doesn't seem meaningful.

If anything maybe a single `let _foo = ...` for the most complex case, with a comment saying you're testing that it can be successfully constructed?